// @scaleflex/ui/core/image-comparison-slider/image-comparison-slider.styles.js
import styled, { css, keyframes } from 'styled-components';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import { Color, Color as PaletteColor } from '../../utils/types/palette';
import { FontVariant } from '../../utils/types/typography';
import { lightPalette } from '../../theme/roots/palette';
var baseClassName = 'image-comparison-slider';
var pulseKeyframe = /*#__PURE__*/keyframes(["0%{opacity:1;}50%{opacity:0.4;}100%{opacity:1;}"]);
var ComparisonSlider = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-111irnx-0"
})(["display:flex;flex-direction:column;gap:10px;user-select:none;"]);
var Image = /*#__PURE__*/styled.img.attrs({
  className: generateClassNames(baseClassName, 'image')
}).withConfig({
  componentId: "sc-111irnx-1"
})(function (_ref) {
  var $isLoading = _ref.$isLoading,
    theme = _ref.theme;
  return css(["width:100%;height:auto;vertical-align:middle;-moz-user-select:none;-webkit-user-select:none;user-select:none;pointer-events:none;touch-action:none;", ""], $isLoading && css(["background-color:", ";animation:", " 1.5s ease-in-out 0.5s infinite;width:100%;height:100%;"], theme === null || theme === void 0 ? void 0 : theme.palette[PaletteColor.ButtonPrimaryText], pulseKeyframe));
});
var LeftImageWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'leftImageWrapper')
}).withConfig({
  componentId: "sc-111irnx-2"
})(["z-index:1;position:absolute;height:100%;width:100%;top:0;overflow:hidden;img{height:100%;object-fit:cover;}"]);
var RightImageWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'rightImageWrapper')
}).withConfig({
  componentId: "sc-111irnx-3"
})(["overflow:hidden;height:100%;width:100%;"]);
var Handle = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'handle')
}).withConfig({
  componentId: "sc-111irnx-4"
})(function (_ref2) {
  var color = _ref2.color,
    _ref2$$thumbIconPaddi = _ref2.$thumbIconPadding,
    $thumbIconPadding = _ref2$$thumbIconPaddi === void 0 ? 10 : _ref2$$thumbIconPaddi,
    _ref2$$thumbIconSize = _ref2.$thumbIconSize,
    $thumbIconSize = _ref2$$thumbIconSize === void 0 ? 10 : _ref2$$thumbIconSize;
  return css(["position:absolute;width:3px;height:100%;background:", ";z-index:3;cursor:col-resize;top:0;svg{display:block;position:absolute;top:", ";right:", ";background:", ";width:", ";height:", ";border-radius:8px;padding:", ";box-sizing:content-box;}"], color, "calc(50% - ".concat($thumbIconPadding + $thumbIconSize / 2, "px)"), "calc(50% - ".concat($thumbIconPadding + $thumbIconSize / 2, "px)"), color, "".concat($thumbIconSize, "px"), "".concat($thumbIconSize, "px"), "".concat($thumbIconPadding, "px"));
});
var FallbackPreviewWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'fallbackPreviewWrapper')
}).withConfig({
  componentId: "sc-111irnx-5"
})(function (_ref3) {
  var backgroundColor = _ref3.backgroundColor,
    gap = _ref3.gap;
  return css(["display:flex;flex-direction:column;gap:", ";width:100%;height:100%;justify-content:center;align-items:center;background:", ";"], "".concat(gap, "px"), backgroundColor);
});
var FallbackPreviewMsg = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'fallbackPreviewMsg')
}).withConfig({
  componentId: "sc-111irnx-6"
})(function (_ref4) {
  var typography = _ref4.theme.typography,
    _ref4$color = _ref4.color,
    color = _ref4$color === void 0 ? lightPalette[Color.TextSecondary] : _ref4$color;
  return css(["", ";user-select:none;color:", ";"], typography.font[FontVariant.TextMedium], color);
});
var SliderWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'sliderWrapper')
}).withConfig({
  componentId: "sc-111irnx-7"
})(["position:relative;overflow:hidden;touch-action:none;"]);
var Footer = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'footer')
}).withConfig({
  componentId: "sc-111irnx-8"
})(function (_ref5) {
  var theme = _ref5.theme;
  return css(["", ";display:flex;width:100%;justify-content:space-between;"], theme.typography.font[FontVariant.TextMedium]);
});
var Styled = applyDisplayNames({
  Image: Image,
  LeftImageWrapper: LeftImageWrapper,
  RightImageWrapper: RightImageWrapper,
  FallbackPreviewMsg: FallbackPreviewMsg,
  Handle: Handle,
  ComparisonSlider: ComparisonSlider,
  FallbackPreviewWrapper: FallbackPreviewWrapper,
  SliderWrapper: SliderWrapper,
  Footer: Footer
});
export default Styled;
